hangman/lib/game/Game.ex

```diff
@@ -0,0 +1,145 @@
+defmodule Hangman.Game do
+  
+  #This is the implement of Hangman.
+  #Here are effects of different:
+  #Game_record: a struct record game 
+  #change_letters: check whether the user's input is correct. if so ,change "_" to user's input 
+  #valid?: used in Enum.reduce, if char =="_", accumulation add 1, else add 0 
+  #count: count numbers of "_" in a string, if the number equals 0, means won
+  #update_record update game's record
+  #play: used to receive and send message, like a simple server to maintain state
+  #deal_empty: if stirng is empty return "cat" else return the string
+  #init_game: initialize a new game with a certain word
+  defmodule Game_record do
+    defstruct game_state: :initializing, turns_left: 7, letters: [], used: [], last_guess: ""
+  end
+  
+  @doc """
+  check whether the user's input is correct. if so ,change "_" to user's input 
+  ##Example
+      iex> Hangman.Game.change_letters(["a","p","p","l","e"],["a","_","_","_","_"],"p")
+      ["a","p","p","_","_"]
+  """
+  def change_letters([],[],_letter), do: []
+  def change_letters([h|t],["_"|t1],h), do: [h|change_letters(t,t1,h)]
+  def change_letters([_h|t],[h|t1],letter), do: [h|change_letters(t,t1,letter)]
+
+  @doc """
+  used in Enum.reduce, if char =="_", accumulation add 1, else add 0 
+  ##Example
+      iex> Hangman.Game.valid?("a",0)
+      0
+      iex> Hangman.Game.valid?("_",0)
+      1
+  """
+  def valid?("_",acc), do: acc+1
+  def valid?(_,acc), do: acc
+  
+  @doc """
+  count numbers of "_" in a string, if the number equals 0, means won
+  ##Example
+      iex> Hangman.Game.count(["a","_","_","l","e"])
+      2
+      iex> Hangman.Game.count(["a","p","p","l","e"])
+      0
+  """
+  def count(letters), do: Enum.reduce(letters,0,&valid?/2)
+
+  @doc """
+  update game's record by user's input
+  if :good_guess or :already_used , truns_left unchanging
+  ##Example
+      iex> Hangman.Game.update_record(["a","p","p","l","e"],%Hangman.Game.Game_record{letters: ["_","_","_","_","_"]},"p")
+      %Hangman.Game.Game_record{game_state: :good_guess, turns_left: 7,letters: ["_","p","p","_","_"], used: ["p"], last_guess: "p"}
+
+      iex> Hangman.Game.update_record(["a","p","p","l","e"],%Hangman.Game.Game_record{letters: ["_","_","_","_","_"]},"x")
+      %Hangman.Game.Game_record{game_state: :bad_guess, turns_left: 6,letters: ["_","_","_","_","_"], used: ["x"], last_guess: "x"}
+  """
+  def update_record(word,%Hangman.Game.Game_record{}=record,letter) do
```

Comment:
## idiomatic -15

All the rest of your code has a really nice functional style. But this function is horrible :)

It is far too big, and is full of conditional logic, making it hard to read and therefore hard to maintain in future.

Whenever you find yourself writing a function that is more than a couple of lines long, make yourself stop. As "what one thing does this function do?". Then write that down in one line of code. To do that, you'll need to assume some new functions exist. So no write those, using the same technique.